### PR TITLE
Use cuda::proclaim_return_type on device lambdas.

### PR DESCRIPTION
**Pull request by bdice**
## Description
This PR makes cudf compatible with Thrust 2 by adding `cuda::proclaim_return_type`, but does not upgrade the version of Thrust/CCCL just yet. Currently we use libcudacxx 2.1.0, which makes `cuda::proclaim_return_type` available, but we still use Thrust 1.17.2 which doesn't require device lambdas to have proclaimed return types. This diff is separated out from #14576 and reduces the diff we must carry in #14576 -- which should contain only packaging changes / version updates for the CCCL 2 migration. This PR is **nonbreaking**, while #14576 will be **breaking**.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/include/cudf/detail/utilities/cast_functor.cuh [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+/**
+ * @brief Functor that casts another functor's result to a specified type.
+ *
+ * CUB 2.0.0 reductions require that the binary operator returns the same type
+ * as the initial value type, so we wrap binary operators with this when used
+ * by CUB.
+ */
+template <typename ResultType, typename F>
+struct cast_functor_fn {
```
Comment (ttnghia) [inline cpp/include/cudf/detail/utilities/cast_functor.cuh:42]: Why do we have this new file?

**Review by ttnghia — APPROVED** [review]

File: cpp/src/copying/copy.cu [hunk 2]
```diff
@@ -35,6 +35,8 @@
 #include <thrust/distance.h>
 #include <thrust/iterator/counting_iterator.h>
 
+#include <cuda/functional>
```
Comment (davidwendt) [inline cpp/src/copying/copy.cu:38]: Doesn't look like this header is needed.

File: cpp/src/io/json/nested_json_gpu.cu [hunk 3]
```diff
@@ -45,6 +45,8 @@
 #include <thrust/transform.h>
 #include <thrust/tuple.h>
 
+#include <cuda/functional>
+
```
Comment (davidwendt) [inline cpp/src/io/json/nested_json_gpu.cu:49]:
```suggestion
```

File: cpp/src/quantiles/tdigest/tdigest_aggregation.cu [hunk 4]
```diff
@@ -623,17 +630,18 @@ std::unique_ptr<column> build_output_column(size_type num_rows,
...
-                    thrust::make_counting_iterator(0) + num_rows,
-                    sizes.begin(),
-                    [offsets = offsets->view().begin<size_type>()] __device__(size_type i) {
-                      return offsets[i + 1] - offsets[i];
-                    });
+  auto offsets_iter = offsets->view().begin<size_type>();
+  thrust::adjacent_difference(rmm::exec_policy(stream),
+                              offsets->view().begin<size_type>() + 1,
+                              offsets->view().end<size_type>(),
+                              sizes.begin());
```
Comment (davidwendt) [inline cpp/src/quantiles/tdigest/tdigest_aggregation.cu:637]: This rewrite assumes that the first offset value is always 0 while the original transform code does not.

**Review by davidwendt — APPROVED** [review]

**Review by vuule — COMMENTED** [review]
never finished the review :(
Wanted to asked about this comment anyway.

File: cpp/include/cudf/column/column_view.hpp [hunk 5]
```diff
@@ -478,7 +478,10 @@ class mutable_column_view : public detail::column_view_base {
  public:
   mutable_column_view() = default;
 
-  ~mutable_column_view() = default;
+  ~mutable_column_view(){
+    // Needed so that the first instance of the implicit destructor for any TU isn't 'constructed'
+    // from a host+device function marking the implicit version also as host+device
```
[the remark this answers is not in the document]
Comment (bdice) [reply cpp/include/cudf/column/column_view.hpp:483]: That's a comment we acquired with help from @robertmaynard. I was suffering from inexplicable compilation problems and this fixed it.

File: cpp/include/cudf/column/column_view.hpp [hunk 6]
```diff
@@ -478,7 +478,10 @@ class mutable_column_view : public detail::column_view_base {
...
+  };
```
Comment (ttnghia) [inline cpp/include/cudf/column/column_view.hpp:484]: This will trigger IDE warning.

```suggestion
  }
```

**Pull request merged**